File: java/weka/classifiers/trees/Splitter.java
package weka.classifiers.trees;

import weka.core.*;
import java.io.Serializable;

/**
 * An abstract class that represents a Splitter object.
 */
abstract class Splitter implements Serializable {

    /**
     * The method to be implemented to find a split based on the given data.
     */
    abstract void findSplit(Instances data);

    /**
     * The method to be implemented that returns true if the given instance
     * should go down the first branch and false otherwise.
     */
    abstract boolean firstSubset(Instance instance);

    /**
     * The method to be implemented that returns the quality of the split.
     */
    abstract double splitQuality();

    /**
     * Returns a string representation of the condition for a branch.
     */
    abstract String toString(boolean leftBranch, int decimalPlaces);
}

File: python/Splitter.py
from abc import ABC, abstractmethod

class Splitter(ABC):
    """
    Abstract base class for a data splitter.
    """

    @abstractmethod
    def find_split(self, X, y, args):
        """
        Finds the best split given the training data.
        Parameters:
            X (np.ndarray): Feature matrix of shape (n_samples, n_features)
            y (np.ndarray): Label vector of shape (n_samples,)
            args: Dictionary with possible arguments
        """
        pass

    @abstractmethod
    def first_subset(self, x):
        """
        Determines if a given instance x belongs to the left (first) subset.
        Parameters:
            x (np.ndarray): Feature vector of shape (n_features,)
        Returns:
            bool: True if instance belongs to left branch, False otherwise
        """
        pass

    @abstractmethod
    def split_quality(self):
        """
        Returns a numeric score for the quality of the current split.
        Returns:
            float
        """
        pass

    @abstractmethod
    def __str__(self, left_branch, decimal_places):
        """
        Returns a string representation of the split condition.
        """
        pass
